### DISCUSSION TITLE
SQLAlchemy 2.0 and greenlet

### DISCUSSION THREAD

**Discussion by agronholm** [sqlalchemy/sqlalchemy#8482] [Usage Questions] [answered]
I'm interested in learning which aspects of SQLAlchemy's design require the use of greenlet. In other words, why can it not be removed in v2.0?

**Comment by CaselIT** [comment]
Hi,

Mike wrote a long post on the python issue tracker when the async support functionality was first added on the 1.4 branch https://github.com/python/cpython/issues/66435#issuecomment-1093662821

The summary of this is basically this:

SQLalchemy is a library that can be considered as in between user code and IO, since it will handle an user code request for a query, will transform it in a sql string, use the db driver to execute the sql (that in most cases involves IO) and return back the result to the user code.

Libraries like this that want to support both sync and asnyc normally have two alternatives:
- have two different implementation for all the code paths that can ultimately use IO 
- write all the code paths that use IO in async format and then offer a wrapper for the sync part (if I'm not mistaken the [psycopg](https://github.com/psycopg/psycopg) driver uses this pattern) 

Using greenlet a 3rd alterative is available
- write all the code paths that use IO in sync format and then use greenlet just on the actual IO calls to switch from a sync greenlet to an async one that can wait for the io.

All the code on sqlalchemy is written in a sync way, and there are many code path that can execute IO, so the first two options basically require either duplicating (and then supporting) a large part of sqlalchemy, or rewriting most of the existing code.
Using the greenlet option instead all the sync code of sqlalchemy can stay as is, most of the changes required are a new async interface used by the user code (that would be required in any case), and dialect implementation that take into consideration that the drivers are async (that's also required in any case).

You can see what I mean by looking at the commit that introduces asyncio support https://github.com/sqlalchemy/sqlalchemy/commit/5fb0138a3220161703e6ab1087319a669d14e7f4. That's a 4k loc change including documentation, the asyncpg dialect and tests on a 600k+ library (including docs and tests) library. Using any of the two other option would require drastically more changes, time and complexity 

Considering the above I don't see the sqlalchemy asyncio implementation without greenlet as something feasible, unless a similar solution is implemented natively by python

**Comment by zzzeek** [comment, marked as answer]
by "removed" do you mean, "run SQLAlchemy without greenlet installed", which is fine as long as you don't use asyncio, or "I want to use SQLAlchemy asyncio without greenlet" ?   There is a theoretical way that would be possible, which is rewrite 100% of SQLAlchemy in asyncio, then provide sync "stubs" - basically turn the whole library inside out.  You'd have to move to an alternate universe to see it actually happen, that effort is many orders of magnitude over the actual benefits.    

There's also to do the thing Django just did and use a threadpool, which IMO is kind of a "cosmetic" version of asyncio that wont give you any real async benefits.

Maybe we will eventually see some pure Python asyncio ORMs that then provide "sync" stubs, which therefore don't need greenlet,  threadpools or two entirely separate implementations to do sync and async.

From my POV though, `async` / `await` are just simple syntactical tricks in Python, and I would prefer if the thing we did using greenlet were simply made a standard part of the Python standard library - basically functional versions of async/await that don't require keywords.

**Reply by agronholm** [reply]
Yeah, I wanted to ask which parts of SQLAlchemy necessitated the use of greenlets instead of coroutine functions. I didn't think the changes required to drop greenlet for async operation would be that pervasive.

**Reply by zzzeek** [reply]
try it sometime!   just to get a connection checked out from the pool is a huge ordeal.   Having it work within the ORM internals, with lazy loading, I would need an additional lifetime to completely rewrite it from scratch, it would be a totally different library.

**Reply by zzzeek** [reply]
it would be vastly less effort for me to instead lobby the Python devs / PEP community to get functional versions of async / await into the standard library, which would mean that nobody ever has to worry about if their library is sync or async ever again as it would be trivial to switch between the two contexts as we can now do in SQLAlchemy.   But as I've said many times, I get the impression that this "separation" is a feature, not a bug, of asyncio.    Hence we use greenlet.
